imap/mailbox.c

```diff
@@ -8236,3 +8235,81 @@ HIDDEN int mailbox_changequotaroot(struct mailbox *mailbox,
   done:
     return r;
 }
+
+struct part_rock {
+    const char *path;
+    unsigned long is_meta;
+    char **partition;
+};
+
+static void get_partition(const char *key, const char *val, void *rock)
+{
+    struct part_rock *prock = (struct part_rock *) rock;
+
+    if (*prock->partition) return;
+
+    if (prock->is_meta) {
+        if (strncmp("meta", key, 4)) return;
+        key += 4;
+    }
+    if (strncmp("partition-", key, 10)) return;
+
+    size_t vlen = strlen(val);
+    if (!strncmp(prock->path, val, vlen) && prock->path[vlen] == '/') {
+        *prock->partition = xstrdup(key+10);
+    }
+}
+
+EXPORTED struct mboxlist_entry *mailbox_mbentry_from_path(const char *path)
+{
+    struct buf buf = BUF_INITIALIZER;
+    size_t pathlen = strlen(path);
+    size_t fnamelen = strlen(FNAME_HEADER);
+    struct mboxlist_entry *mbentry = mboxlist_entry_create();
+    struct mailbox mailbox;
+    int r;
+
+    if (pathlen < fnamelen || strcmp(path + pathlen - fnamelen, FNAME_HEADER)) {
+        buf_setcstr(&buf, path);
+        buf_appendcstr(&buf, FNAME_HEADER);
+        path = buf_cstring(&buf);
+    }
+
+    zeromailbox(mailbox);
+    mailbox.mbentry = mbentry;
+
+    r = mailbox_read_header(&mailbox, path);
+    if (!r) {
+        mbentry->mbtype = mailbox.h.mbtype;
+
+        /* Steal strings from mailbox_header */
+        mbentry->name = mailbox.h.name;
+        mbentry->uniqueid = mailbox.h.uniqueid;
```

Comment:
uniqueid might be NULL here, if the header file was missing that field for whatever reason

If `path` is a UUID directory, we should probably get the uniqueid from the path rather than from the header contents, otherwise we might have to rename the directory later instead?

Reply:
Done.